### PR TITLE
[WIP] LLAMA3 Integration with BioNeMo Recipes with ASCII Tokenization/EVO2-Style Dataloader

### PR DESCRIPTION
**Pull request by savitha-eng**
### Description
This PR implements a complete end-to-end genomic foundation model training pipeline using LLAMA3 architecture with EVO2-compatible data processing and distributed training capabilities.
#### Key Components:
**Genomic Tokenization System:**
ASCII character-level tokenization with 256-token vocabulary covering complete ASCII character set
Direct nucleotide encoding supporting standard nucleotides (A, T, C, G, N) and IUPAC ambiguity codes
EVO2-compatible byte-level tokenization methodology
Integrated BOS, EOS, and PAD tokens for proper sequence boundary management
**EVO2-Style Genomic Dataloader (following the Sharded Eden Dataloader):**
Efficient sequence windowing: 8192-token sequences with 7992-token stride (200bp overlap) - following the EVO2 approach
Randomization: fixed window positions with shuffled access order for training stability (
SQLite backend integration for high-performance sequence retrieval
Full FSDP and DDP distributed training compatibility
Memory-efficient streaming with configurable parameters
Currently processes small subsample dataset (1,024 sequences into 254,043 training windows) with future scaling planned for production datasets
**Model Pipeline Architecture:**
Dynamic RoPE implementation supporting arbitrary sequence lengths with ESM-2 style computation
Random weight initialization using from-config approach to avoid text-domain bias
Memory-optimized LLAMA3: 8-layer, 403M parameter configuration for development/testing
FSDP integration with TransformerEngine optimization
Complete Weights & Biases integration for experiment tracking
**Quick Validation Results: - Much more testing needed**
L0 Sanity: 4-layer model, 25 steps, 3-minute validation
L1 Pilot: 8-layer model, 200 steps, 50-minute training
Stable training convergence demonstrated (loss: 6.1 → 1.38 over 200 steps)
Consistent 23.5GB GPU memory usage with 4.1 iterations/second sustained performance
Note: Much more extensive validation is needed with full-scale models and production datasets
### Usage
Quick validation run:
`cd bionemo-recipes/recipes/llama_native_te_nvfsdp
torchrun --nproc_per_node=1 train.py --config-name=L0_sanity`
More Realistic training run (still very small scale):
`cd bionemo-recipes/recipes/llama_native_te_nvfsdp
torchrun --nproc_per_node=1 train.py --config-name=L1_pilot`
Configuration customization:
Type of changes
[x] New feature (non-breaking change which adds functionality)
[ ] Bug fix (non-breaking change which fixes an issue)
[ ] Refactor
[ ] Documentation update
[ ] Other (please describe):
CI Pipeline Configuration
Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.
[ ] ciflow:skip - Skip all CI tests for this PR
[ ] ciflow:notebooks - Run Jupyter notebooks execution tests for bionemo2
[ ] ciflow:slow - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
[ ] ciflow:all - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
[x] ciflow:all-recipes - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.
Unit tests marked as @pytest.mark.multi_gpu or @pytest.mark.distributed are not run in the PR pipeline.
For more details, see CONTRIBUTING
> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.
Authorizing CI Runs
We use copy-pr-bot to manage authorization of CI
runs on NVIDIA's compute resources.
If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
/ok to test comment on the pull request to trigger CI. This will need to be done for each new commit.
Pre-submit Checklist
[x] I have tested these changes locally
[ ] I have updated the documentation accordingly
[ ] I have added/updated tests as needed
[] All existing tests pass successfully
### Next Steps:
- Dataset Scaling: Scale from current subsample (1K sequences) to larger datasets (10M+ sequences)
- Model Scaling: Increase to 16+ layers and full parameter counts for training
- Multi-Node Training: Test and validate existing FSDP implementation across multiple GPU nodes
- THD Sequence Packing
- Extended Context Training: Support for longer context lengths (32K+ tokens) with iterative training strategies and memory optimization
- Extended Validation: Comprehensive evaluation with downstream genomic tasks and comparative benchmarking against EVO2

### CODE REVIEW

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/__init__.py
Comment (pstjohn) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/__init__.py:1]: this folder isn't a package, so this file won't do anything 🤷

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+from typing import Dict, Optional, Tuple, Iterator
+from pathlib import Path
+import torch
+from torch.utils.data import Dataset, DataLoader, DistributedSampler
+from omegaconf import DictConfig
+
+from tokenizer import NucleotideASCIITokenizer
+
+
+class GenomicSequenceDataset(Dataset):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:26]: This dataset class isn't quite right; high-level if we want to be consistent with the ESM-2 recipe (and generally follow the huggingface flow) we'd want a dataset that doesn't do tokenization, it just returns the full sequences. I'm not seeing a reason that couldn't essentially be a bunch of sharded parquet files with the raw sequences in them, like we do for esm-2:

https://huggingface.co/datasets/nvidia/esm2_uniref_pretraining_data

then you'd want various map operations to tokenize the dataset from there, and possibly a sampler to shuffle it.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            raise FileNotFoundError(f"Database not found: {self.database_path}")
+            
+        with sqlite3.connect(str(self.database_path)) as conn:
+            cursor = conn.cursor()
+            cursor.execute("SELECT contig_id, length FROM sequences ORDER BY contig_id")
+            self.sequences = cursor.fetchall()
+            
+        print(f"Loaded {len(self.sequences)} sequences from database")
+    
+    def _create_window_mappings(self):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:97]: why does the `stride` parameter from hf tokenizers not do this?
https://huggingface.co/docs/transformers/en/main_classes/tokenizer

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            # Create window mappings for this sequence (fixed positions)
+            for i, start_pos in enumerate(positions):
+                effective_length = min(self.seq_length, seq_length - start_pos)
+                if effective_length >= self.min_window_length:
+                    self.window_mappings.append((seq_idx, contig_id, start_pos, effective_length))
+                    total_windows += 1
+        
+        print(f"Created {total_windows:,} window mappings with stride={self.stride} (overlap={self.seq_length - self.stride}bp)")
+    
+    def _randomize_window_access(self):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:129]: this won't scale very well :D this just shuffles the dataset though?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        1. Map shuffled index to actual window (randomized lookup)
+        2. Retrieve sequence from SQL database (tiled position)
+        3. Tokenize with ASCII encoding (then tokenization)
+        """
+        # Map shuffled index to actual window (randomized lookup)
+        actual_idx = self.shuffled_indices[idx % len(self.shuffled_indices)]
+        seq_idx, contig_id, start_pos, window_length = self.window_mappings[actual_idx]
+        
+        # Retrieve sequence from SQL database (speed requirement)
+        with sqlite3.connect(str(self.database_path)) as conn:
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:159]: yeah, we used sqlite for esm2 in bionemo2 for various reasons, but this really shouldn't be our standard approach. this won't scale very well

**Review by pstjohn — COMMENTED** [review]
leaving a first set of comments, will keep reviewing later

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/__init__.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/__init__.py:4]: testing

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        rc_augmentation: bool = False,
+    ):
+        """
+        Initialize genomic dataset with EVO2-style tiling.
+        
+        Args:
+            database_path: Path to SQLite database with sequences
+            seq_length: Window size (8192 for LLAMA3)
+            tokenizer: ASCII tokenizer for nucleotides
+            stride: Stride between windows (7992 = 8192-200 for EVO2 overlap)
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:53]: Nit: Can we get some sort of explanation why stride is required?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            self.tokenizer = NucleotideASCIITokenizer()
+        else:
+            self.tokenizer = tokenizer
+            
+        # Initialize random state
+        self.rng = np.random.RandomState(seed)
+        random.seed(seed)
+        
+        # Step 1: Load sequences from database
+        self._load_sequences()
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:75]: What are the memory requirements for this?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        
+        # Step 2: First create window mappings (EVO2 style)
+        self._create_window_mappings()
+        
+        # Step 3: Randomize lookups - shuffle window indices
+        self._randomize_window_access()
+        
+        print(f"Dataset ready: {len(self.window_mappings):,} windows with EVO2-style tiling")
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:83]: use logger over print

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    
+    def _load_sequences(self):
+        """Load sequence metadata from SQLite database."""
+        if not self.database_path.exists():
+            raise FileNotFoundError(f"Database not found: {self.database_path}")
+            
+        with sqlite3.connect(str(self.database_path)) as conn:
+            cursor = conn.cursor()
+            cursor.execute("SELECT contig_id, length FROM sequences ORDER BY contig_id")
+            self.sequences = cursor.fetchall()
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:93]: so you're loading the entire sqlite into memory? If that sqlite grows you will go OOM

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            
+        print(f"Loaded {len(self.sequences)} sequences from database")
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:95]: logger > print

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py (hunk 3 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:129]: Can we use a shuffle buffer instead?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        This ensures we don't just get position-0 windows from each sequence.
+        """
+        # Create shuffled indices for randomized access
+        self.shuffled_indices = list(range(len(self.window_mappings)))
+        random.shuffle(self.shuffled_indices)
+        print(f"Shuffled {len(self.shuffled_indices):,} window indices for randomized access")
+    
+    def __len__(self) -> int:
+        """Return number of windows."""
+        return len(self.window_mappings)
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:143]: I would think that the length of the dataset would be the number of rows in the SQlite, but here I see its the number of window mappings?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py (hunk 4 above)
Comment (savitha-eng) [reply bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:159]: @pstjohn  what would you suggest instead for better scaling? I used it because it was there in sharded_eden_dataloader.py (which I was following as a model). If there's another dataloader that I can look at as an example please let me know.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        # Create tensors
+        tokens = torch.tensor(token_ids, dtype=torch.long)
+        
+        # Create labels for autoregressive training
+        labels = tokens.clone()
+        labels[:-1] = tokens[1:]
+        labels[-1] = self.tokenizer.pad_token_id
+        
+        # Create attention mask (1 for real tokens, 0 for padding)
+        attention_mask = (tokens != self.tokenizer.pad_token_id).long()
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:194]: I thought this would be something like 
```
attention_mask=self.tokenizer['attention_mask']
```
Is that not the case?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        
+        # Create loss mask (don't compute loss on special tokens)
+        loss_mask = attention_mask.clone()
+        loss_mask[tokens == self.tokenizer.bos_token_id] = 0
+        loss_mask[tokens == self.tokenizer.eos_token_id] = 0
+        
+        return {
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:201]: Since you're casting seq_idx to your data type, do you want to cast the rest of the things to a specific data type here?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    """
+    Create genomic dataloader compatible with Bruno's training script.
+    
+    This creates an infinite iterator matching Bruno's interface while
+    implementing EVO2-style tiling and randomization.
+    """
+    # Create simple dataset for Bruno's training loop
+    dataset = GenomicSequenceDataset(
+        database_path=args.dataset.database_path,
+        seq_length=args.dataset.seq_length,
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:243]: is `seq_length` --> `max_seq_len`? Since if you have sequences < seq_len you're going to pad them right?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        tokenizer=tokenizer,
+        stride=args.dataset.get("stride", args.dataset.seq_length - 200),  # EVO2 default
+        min_window_length=args.dataset.get("min_window_length", 1000),
+        seed=args.dataset.get("seed", 42),
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:247]: seed should come from a config indeally

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+        num_workers=args.dataset.get("num_workers", 0),
+        pin_memory=True,
+        drop_last=True,
+        persistent_workers=args.dataset.get("num_workers", 0) > 0,
+    )
+    
+    # Calculate epoch length  
+    epoch_len = len(dataloader)  # Use dataloader length for distributed setting
+    
+    print(f"Created genomic dataloader: {len(dataset):,} windows, {epoch_len:,} batches per epoch")
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:268]: logger > print please.
logger.info(f"blahblah {<var>}")

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py (hunk 2 above)
Comment (jstjohn) [reply bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:97]: I think there are two potential issues:
1. Shuffling would happen at the dataset level, so you would randomly iterate through input sequences, but then each of those input sequences would get mapped to many ordered windows by the tokenizer. As @pstjohn suggested in the slack thread, this could be helped with some kind of a local random buffer. Depending on the size of the buffer and how many chunks are returned from a given genome sequence it may take a while before you eventually start sampling from different sequences.
2. You would need to handle the fact that your dataset returns N items and your tokenizer applied to the dataset returns M >>= N items given the 1-many relationship of the tokenizer to the inputs to the tokenizer if you went this path. If you pass `stride, max_length, and return_overflowing_tokens=True` to your tokenizer call wherever that happens then that would return ordered strided window samplings for a given input which you would then want to reshuffle and repack into smaller batches of a target size ideally.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py (hunk 9 above)
Comment (jstjohn) [reply bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py:93]: I think it's just the id,length tuple, not the full sequences, so it doesn't OOM in practice on a 1T database.
